views/dashboard.ejs

```diff
@@ -0,0 +1,293 @@
+            }
+          }
+        }
+      }
+
+    }
+
+    async function fetchDataValidationItems() {
+      const data = await fetch(window.location.href + '/etl-validation')
+        .catch(e => console.log(e));
+      return await data === undefined ? '' : data.json().catch(e => console.log(e));
+    }
+
+    function renderValidationItems(items) {
+      if (items == null) {
+        return;
+      }
+      let div = document.getElementById('etl-validation');
+      div.innerHTML =
+        items.map(L => {
+          return '<tr><td style="color: navy; white-space: nowrap; border: none; padding:4">' + L.test + '</td><td style="color:blue; border: none; padding:4;">' + L.details + '</td><td style="color:blue; border: none; padding:4;">' + L.id + '</td><td style="color:blue; border: none; padding:4;">' + L.listing + '</td></tr>'
+        }).join('')
+    }
+
+    async function staging() {
+
+      /* Prompt for confirmation */
+      let confirmation = confirm("Import data to preview site?");
+      if (confirmation == false) { return; }
+      document.getElementById('status').innerText = 'Preparing to run import process...';
+
+      /* Safety check to abort if ETL is in progress */
+      const logs = await fetchEtlLogs();
+      const hasStartMessage = (logs.filter(i => i.message === 'Python ETL Script Start')).length >= 1;
+      const hasEndMessage = (logs.filter(i => i.message === 'Python ETL Script End')).length >= 1;
+      if (hasStartMessage && !hasEndMessage) {
+        document.getElementById('status').innerText = 'There is already an import in progress. Please allow up to 30 minutes for that process to complete.';
+      }
+
+      /* Trigger the ETL process to start */
+      fetch(window.location.href + '?' + new URLSearchParams({ action: 'runetl' }), { method: 'GET' })
+        .catch(e => { document.getElementById('status').innerText = 'Failed to start import process'; })
+
+      /* Query the server for ETL status on an interval */
+      var stagingInterval = window.setInterval(async function () {
+
+        /* Fetch and display ETL logs */
+        const logs = await fetchEtlLogs();
+        renderEtlLogs(logs);
+
+        /* Update the status message */
+        const jobStart = logs.filter(L => L.message === 'Job Start');
+        const jobEnd = logs.filter(L => L.message === 'Python ETL Script End');
+        const stagingFinalized = logs.filter(L => L.message === 'Finalize the staging table');
+        if (stagingFinalized && stagingFinalized.length >= 1) {
+          const items = await fetchDataValidationItems();
+          renderValidationItems(items)
+        }
+        if (jobStart && jobStart.length >= 1) {
+          document.getElementById('status').innerHTML = 'Import process is running...';
+        }
+        if (jobEnd && jobEnd.length >= 1) {
+          const previewPage = [window.location.protocol, '//', window.location.host.replace(/\d+/, '3000'), '?datatable=staging'].join('');
```

Comment:
could be a good opportunity for `query-string` but just a suggestion.

Reply:
This code is tricky, since it is meant to work both for local and hosted environments. It needs to replace the port number if one exists. query-string just parses a given URL, if I'm looking at the documentation correctly.

Reply:
@kentpmckinney Yes you are totally right.  Hope I didn't send you down a rabbit hole.